src/pkg/sources/tarball.go

```diff
@@ -0,0 +1,243 @@
+		return err
+	}
+
+	var zarfYamlSHA string
+	for _, layer := range imageManifest.Layers {
+		if layer.Annotations[ocispec.AnnotationTitle] == config.ZarfYAML {
+			zarfYamlSHA = layer.Digest.Encoded()
+			break
+		}
+	}
+
+	if zarfYamlSHA == "" {
+		return fmt.Errorf(fmt.Sprintf("zarf.yaml with SHA %s not found", zarfYamlSHA))
+	}
+
+	// reset file reader
+	_, err = sourceArchive.Seek(0, 0)
+	if err != nil {
+		return err
+	}
+	if err := format.Extract(ctx, sourceArchive, []string{filepath.Join(config.BlobsDir, zarfYamlSHA)}, func(ctx context.Context, fileInArchive av4.File) error {
+		// write zarf.yaml to tmp for checking optional components later on
+		dst := filepath.Join(dst.Base, config.ZarfYAML)
+		outFile, err := os.Create(dst)
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+		stream, err := fileInArchive.Open()
+		if err != nil {
+			return err
+		}
+		defer stream.Close()
+		_, err = io.Copy(outFile, io.Reader(stream))
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		err = sourceArchive.Close()
+		if err != nil {
+			return err
+		}
+		return err
+	}
+
+	err = sourceArchive.Close()
+	return err
+}
+
+// Collect doesn't need to be implemented
+func (t *TarballBundle) Collect(_ string) (string, error) {
+	return "", nil
+}
+
+// extractPkgFromBundle extracts a Zarf package from a local tarball bundle
+func (t *TarballBundle) extractPkgFromBundle() error {
+	ctx := context.TODO()
+	format := av4.CompressedArchive{
+		Compression: av4.Zstd{},
+		Archival:    av4.Tar{},
+	}
+
+	sourceArchive, err := os.Open(t.BundleLocation)
+	if err != nil {
+		return err
+	}
+
+	var manifest oci.ZarfOCIManifest
+	if err := format.Extract(ctx, sourceArchive, []string{filepath.Join(config.BlobsDir, t.PkgManifestSHA)}, utils.ExtractJSON(&manifest)); err != nil {
+		if err := sourceArchive.Close(); err != nil {
+			return err
+		}
+		return err
+	}
+
+	if err := sourceArchive.Close(); err != nil {
+		return err
+	}
+
+	extractLayer := func(_ context.Context, file av4.File) error {
+		if file.IsDir() {
+			return nil
+		}
+		stream, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer stream.Close()
+
+		desc := helpers.Find(manifest.Layers, func(layer ocispec.Descriptor) bool {
+			return layer.Digest.Encoded() == filepath.Base(file.NameInArchive)
+		})
+
+		path := desc.Annotations[ocispec.AnnotationTitle]
+		size := desc.Size
+		layerDst := filepath.Join(t.TmpDir, path)
+		if err := zarfUtils.CreateDirectory(filepath.Dir(layerDst), 0700); err != nil {
+			return err
+		}
+
+		target, err := os.Create(layerDst)
+		if err != nil {
+			return err
+		}
+		defer target.Close()
+
+		written, err := io.Copy(target, stream)
+		if err != nil {
+			return err
+		}
+		if written != size {
+			return fmt.Errorf("expected to write %d bytes to %s, wrote %d", size, path, written)
+		}
+
+		return nil
+	}
```

Comment:
just like how `sources.TarballSource` works, this is where a filelist should be built, then later fed into `dst.SetFromPaths` as this is the true source of where a file placed on disk is from. Extracting each file by name, then doing a `filepath.Walk` to get the names again is a wasted operation.

Reply:
Fo sho, good call, removed Walk